services/venice-server/src/main/java/com/linkedin/venice/listener/ReadQuotaEnforcementHandler.java

```diff
@@ -330,27 +307,77 @@ private void updateQuota(PartitionAssignment partitionAssignment) {
-      if (v == null || v.getAmortizedRefillPerSecond() * enforcementIntervalSeconds != newRefillAmount) {
-        // only replace the existing bucket if the difference is greater than 1
-        return tokenBucketfromRcuPerSecond(quotaInRcu, thisNodeQuotaResponsibility);
-      } else {
-        return v;
-      }
-    });
-    String storeName = Version.parseStoreFromVersionTopic(topic);
-    stats.setStoreTokenBucket(storeName, getBucketForStore(storeName));
+    storeVersionRateLimiters.compute(
+        topic,
+        (k, v) -> getRateLimiter(
+            topic,
+            quotaInRcu,
+            thisNodeQuotaResponsibility,
+            v,
+            storeVersionRateLimiterType,
+            quotaEnforcementIntervalInMs,
+            enforcementCapacityMultiple,
+            clock));
   }
 
-  private long calculateRefillAmount(long totalRcuPerSecond, double thisBucketProportionOfTotalRcu) {
-    long totalRefillAmount = totalRcuPerSecond * enforcementIntervalSeconds;
-    return (long) Math.ceil(totalRefillAmount * thisBucketProportionOfTotalRcu);
+  /**
+   * Get the rate limiter for the store version. If the rate limiter is already created and the quota is the same,
+   * return the existing rate limiter.
+   */
+  static VeniceRateLimiter getRateLimiter(
+      String storeVersionName,
+      long quotaInRcu,
+      double thisNodeQuotaResponsibility,
+      VeniceRateLimiter currentRateLimiter,
+      RateLimiterType rateLimiterType,
+      int quotaEnforcementIntervalInMs,
+      int enforcementCapacityMultiple,
+      Clock clock) {
+    // If the rate limiter is already created and the quota is the same, return the existing rate limiter
+    long newQuota = (long) Math.ceil(quotaInRcu * thisNodeQuotaResponsibility);
+    if (currentRateLimiter != null && currentRateLimiter.getGlobalPermitsCount() == quotaInRcu
+        && currentRateLimiter.getThisNodeResponsibility() == thisNodeQuotaResponsibility) {
+      return currentRateLimiter;
+    }
+
+    VeniceRateLimiter newRateLimiter;
+    if (rateLimiterType == RateLimiterType.EVENT_THROTTLER_WITH_SILENT_REJECTION) {
+      LOGGER.info(
+          "Creating EventThrottler rate limiter for store version: {} with quota: {}",
+          storeVersionName,
+          newQuota);
+      newRateLimiter =
+          new EventThrottler(newQuota, quotaEnforcementIntervalInMs, storeVersionName, true, REJECT_STRATEGY);
+    } else if (rateLimiterType == RateLimiterType.GUAVA_RATE_LIMITER) {
+      LOGGER.info(
+          "Creating GuavaRateLimiter rate limiter for store version: {} with quota: {}",
+          storeVersionName,
+          newQuota);
+      newRateLimiter = new GuavaRateLimiter(newQuota);
+    } else {
+      LOGGER.info(
+          "Creating TokenBucket rate limiter for store version: {} with quota: {} and thisNodeQuotaResponsibility: {} quotaEnforcementIntervalInMs: {} enforcementCapacityMultiple: {}",
```

Comment:
Similar here, if we want to log about these specific configs regarding to token bucket type we can log it during initialization of `ReadQuotaEnforcementHandler` since these configs are also not changing from store to store.